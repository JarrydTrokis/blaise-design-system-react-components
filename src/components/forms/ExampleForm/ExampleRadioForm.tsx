import React, { useState } from "react";
import { validateRadio } from "./FormValidation";
import StyledForm, { FormFieldObject } from "../StyledForm";

function ExampleForm() {
    const [formStatus, setFormStatus] = useState<string>("");

    /** List of fields in order for form generation */
    const formElements: FormFieldObject[] = [
        {
            name: "topping",
            description: "Select your favorite topping",
            type: "radio",
            validate: validateRadio,
            radioOptions: [
                { id: "bacon", value: "bacon", label: "Bacon" },
                {
                    id: "cheese", value: "cheese", label: "Cheese", description: "This includes all types of cheese",
                },
            ],
        },
        {
            name: "options",
            description: "Select a radio option",
            type: "radio",
            radioOptions: [
                { id: "firstOption", value: "firstOption", label: "First Option" },
                {
                    id: "other",
                    value: "other",
                    label: "other",
                    specifyOption: {
                        id: "other-text", name: "other-text", description: "Please specify", type: "text",
                    },
                },
            ],
        },
    ];

    /**
     * Function is called after submit of form and all field validation is valid
     *
     * @param formValues Object with all field values
     * @param setSubmitting Function to set isSubmitting attribute which disables submit button while processing the form
     */
    function onFormSubmission(formValues: any, setSubmitting: (isSubmitting: boolean) => void): void {
        console.warn(formValues);
        setFormStatus(`Form submitted, topping chosen ${formValues.topping}`);
        setSubmitting(false);
    }

    return (
        <>
            <p>{formStatus}</p>
            <StyledForm fields={formElements} onSubmitFunction={onFormSubmission} />
        </>
    );
}

export default ExampleForm;
