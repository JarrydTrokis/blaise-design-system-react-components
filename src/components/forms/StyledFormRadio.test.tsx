import React from 'react';
import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import ExampleRadioForm from "./ExampleForm/ExampleRadioForm";

test('it matches snapshots',  async () => {
    const formPage = render(<ExampleRadioForm/>);

    expect(formPage).toMatchSnapshot();
});

test('it displays the the radio option description',  async () => {
    render(<ExampleRadioForm/>);

    await waitFor(() => {
        const errorMessage = screen.getByText(/This includes all types of cheese/i);
        expect(errorMessage).toBeInTheDocument();

    });
});


test('error appears on submit of empty form',  async () => {
    render(<ExampleRadioForm/>);

    const submitButton = screen.getByTestId(/submit-button/i);
    fireEvent.click(submitButton)

    await waitFor(() => {
        const errorMessage = screen.getByText(/There is 1 problem with your answer/i);
        expect(errorMessage).toBeInTheDocument();

        const radioErrorMessage = screen.queryAllByText(/Select an option/i);
        expect(radioErrorMessage).toHaveLength(2);

    });
});


test('submit function is called when form is valid', async () => {
    render(<ExampleRadioForm/>);

    fireEvent.click(screen.getByLabelText(/Bacon/i), );

    const submitButton = screen.getByTestId(/submit-button/i);
    fireEvent.click(submitButton)

    await waitFor(() => {
        const successMessage = screen.getByText(/Form submitted, topping chosen bacon/i);
        expect(successMessage).toBeInTheDocument()
    });
});
