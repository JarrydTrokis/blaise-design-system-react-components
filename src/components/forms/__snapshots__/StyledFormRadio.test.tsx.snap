// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshots 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p />
      <form
        action="#"
      >
        <fieldset
          class="ons-fieldset"
        >
          <legend
            class="ons-fieldset__legend"
          >
            Select your favorite topping
          </legend>
          <div
            class="ons-radios__items"
            id="topping"
          >
            <p
              class="ons-radios__item"
            >
              <span
                class="ons-radio"
              >
                <input
                  class="ons-radio__input js-radio"
                  id="bacon"
                  name="topping"
                  type="radio"
                  value="bacon"
                />
                <label
                  class="ons-radio__label "
                  for="bacon"
                  id="bacon-label"
                >
                  Bacon
                </label>
              </span>
            </p>
            <br />
            <p
              class="ons-radios__item"
            >
              <span
                class="ons-radio"
              >
                <input
                  class="ons-radio__input js-radio"
                  id="cheese"
                  name="topping"
                  type="radio"
                  value="cheese"
                />
                <label
                  class="ons-radio__label ons-label--with-description"
                  for="cheese"
                  id="cheese-label"
                >
                  Cheese
                  <span
                    class="ons-label__description ons-radio__label--with-description"
                    id="white-label-description-hint"
                  >
                    This includes all types of cheese
                  </span>
                </label>
              </span>
            </p>
            <br />
          </div>
        </fieldset>
        <fieldset
          class="ons-fieldset"
        >
          <legend
            class="ons-fieldset__legend"
          >
            Select a radio option
          </legend>
          <div
            class="ons-radios__items"
            id="options"
          >
            <p
              class="ons-radios__item"
            >
              <span
                class="ons-radio"
              >
                <input
                  class="ons-radio__input js-radio"
                  id="firstOption"
                  name="options"
                  type="radio"
                  value="firstOption"
                />
                <label
                  class="ons-radio__label "
                  for="firstOption"
                  id="firstOption-label"
                >
                  First Option
                </label>
              </span>
            </p>
            <br />
            <p
              class="ons-radios__item"
            >
              <span
                class="ons-radio"
              >
                <input
                  class="ons-radio__input js-radio"
                  id="other"
                  name="options"
                  type="radio"
                  value="other"
                />
                <label
                  class="ons-radio__label "
                  for="other"
                  id="other-label"
                >
                  other
                </label>
                <span
                  class="ons-radio__other ons-radio__other--open"
                  id="other-radio-other-wrap"
                >
                  <label
                    class="ons-label ons-u-fs-s--b "
                    for="other-text"
                    id="other-textbox-label"
                  >
                    Please specify
                  </label>
                  <input
                    class="ons-input ons-input--text ons-input-type__input ons-input--w-auto"
                    id="other-text"
                    name="other-text"
                    type="text"
                    value=""
                  />
                </span>
              </span>
            </p>
            <br />
          </div>
        </fieldset>
        <br />
        <button
          class="ons-btn "
          data-testid="submit-button"
          type="submit"
        >
          <span
            class="ons-btn__inner"
          >
            Save and continue
          </span>
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <p />
    <form
      action="#"
    >
      <fieldset
        class="ons-fieldset"
      >
        <legend
          class="ons-fieldset__legend"
        >
          Select your favorite topping
        </legend>
        <div
          class="ons-radios__items"
          id="topping"
        >
          <p
            class="ons-radios__item"
          >
            <span
              class="ons-radio"
            >
              <input
                class="ons-radio__input js-radio"
                id="bacon"
                name="topping"
                type="radio"
                value="bacon"
              />
              <label
                class="ons-radio__label "
                for="bacon"
                id="bacon-label"
              >
                Bacon
              </label>
            </span>
          </p>
          <br />
          <p
            class="ons-radios__item"
          >
            <span
              class="ons-radio"
            >
              <input
                class="ons-radio__input js-radio"
                id="cheese"
                name="topping"
                type="radio"
                value="cheese"
              />
              <label
                class="ons-radio__label ons-label--with-description"
                for="cheese"
                id="cheese-label"
              >
                Cheese
                <span
                  class="ons-label__description ons-radio__label--with-description"
                  id="white-label-description-hint"
                >
                  This includes all types of cheese
                </span>
              </label>
            </span>
          </p>
          <br />
        </div>
      </fieldset>
      <fieldset
        class="ons-fieldset"
      >
        <legend
          class="ons-fieldset__legend"
        >
          Select a radio option
        </legend>
        <div
          class="ons-radios__items"
          id="options"
        >
          <p
            class="ons-radios__item"
          >
            <span
              class="ons-radio"
            >
              <input
                class="ons-radio__input js-radio"
                id="firstOption"
                name="options"
                type="radio"
                value="firstOption"
              />
              <label
                class="ons-radio__label "
                for="firstOption"
                id="firstOption-label"
              >
                First Option
              </label>
            </span>
          </p>
          <br />
          <p
            class="ons-radios__item"
          >
            <span
              class="ons-radio"
            >
              <input
                class="ons-radio__input js-radio"
                id="other"
                name="options"
                type="radio"
                value="other"
              />
              <label
                class="ons-radio__label "
                for="other"
                id="other-label"
              >
                other
              </label>
              <span
                class="ons-radio__other ons-radio__other--open"
                id="other-radio-other-wrap"
              >
                <label
                  class="ons-label ons-u-fs-s--b "
                  for="other-text"
                  id="other-textbox-label"
                >
                  Please specify
                </label>
                <input
                  class="ons-input ons-input--text ons-input-type__input ons-input--w-auto"
                  id="other-text"
                  name="other-text"
                  type="text"
                  value=""
                />
              </span>
            </span>
          </p>
          <br />
        </div>
      </fieldset>
      <br />
      <button
        class="ons-btn "
        data-testid="submit-button"
        type="submit"
      >
        <span
          class="ons-btn__inner"
        >
          Save and continue
        </span>
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
